### ISSUE TITLE
THREE.AnimationAction: this._mixer is undefined

### ISSUE THREAD

**Issue by weiserhei** [mrdoob/three.js#10326]
##### Description of the problem 
Calling a reference to an ```THREE.AnimationAction``` fails with ```TypeError: this._mixer is undefined```.

Im loading a json model, setting up the animations and some time later i want to play the animations.
Im attaching the AnimationActions to my Object so i can play them later via ```myObject.playMe();```. This fails with a type error, while wrapping it inside an anonymous function works:

    function MyClass() {
        //...
	    this.playMe1 = mixer.clipAction( mesh.geometry.animations[ 1 ] ).play; 
	    this.playMe2 = function() { mixer.clipAction( mesh.geometry.animations[ 1 ] ).play() };
    }
    
    var myObjet = new MyClass();
    myObject.playMe1(); // TypeError: this._mixer is undefined
    myObject.playMe2(); // works

##### Three.js version

- [x] Dev
- [x] r82

**Comment by mrdoob** [comment]
Do you mind creating a `jsfiddle`?

**Comment by weiserhei** [comment]
Replace model path, a click-Event triggers: http://jsfiddle.net/f17Lz5ux/5358/

**Comment by mrdumb** [comment]
Your "playMe1" approach is wrong.
PLAY is a function, not a variable.

**Comment by mrdumb** [comment]
function MyClass() {
	    //...
	    this.playMe1 = mixer.clipAction( mesh.geometry.animations[ 1 ] ); 
	    this.playMe2 = function() { mixer.clipAction( mesh.geometry.animations[ 1 ] ).play() };
	}
	
	var myObjet = new MyClass();
	myObject.playMe1.play(); // works
	myObject.playMe2(); // works

**Comment by weiserhei** [comment]
The question is why cant I store a reference to ```play()``` in my attribute and access it? Not storing the ```AnimationAction``` in the attribute and calling ```play()``` on that. Its nonobvious to me why this results in a ```_mixer undefined``` error.

**Comment by mrdumb** [comment]
> The question is why cant I store a reference to play() in my attribute and access it?

You can.
See "playMe2".

**Comment by weiserhei** [comment]
This not the same and I would expect it to work:
```this.playMe1 = animation.play;``` 
> while wrapping it inside an anonymous function works
```this.playMe2 = function() { animation.play(); }```

**Comment by mrdumb** [comment]
> I would expect it to work:
> this.playMe1 = animation.play;

No, it doesn't.
For the last time: PLAY is a function, not a variable.

**Comment by weiserhei** [comment]
I'm not exactly sure what you mean, so let me give an example. I think you're saying that you can't assign a function to a variable, so here's an example showing how to do that: https://jsfiddle.net/5z58y7tz/1/ 

If that's not what you're talking about, can you describe what you mean in a bit more detail?

**Comment by mrdumb** [comment]
> I'm not exactly sure what you mean

PLAY is a function and your "playMe2" is also a function.
That's OK and it works.
PLAY is a function and your ""playMe1" is a variable.
That's wrong !

Clear enough now?

**Comment by knod** [comment]
From what I can see, weirserhei, there should be no error. I'm not even sure how that error would exist. I think I've created an example that more closely resembles the situation you first described: https://jsfiddle.net/4dag9ny6/.

I'm not sure why the code you wrote at the top of this thread should behave differently, but maybe I'm taking crazy pills. If that code is in some way fundamentally different than the example I posted, a new example showing the difference between my fiddle and the code at the top of the thread could help clear up what's going on here.

**Comment by mrdumb** [comment]
> I think I've created an example that more closely resembles the situation you first described

Not true.
In your example "playMe1" is also a function.

**Comment by knod** [comment]
From my example: `this.playMe1 = gotMethods(  "Foosball's da bomb. 1."  ).play;`
From the example at the top: `this.playMe1 = mixer.clipAction( mesh.geometry.animations[ 1 ] ).play`

**Comment by mrdumb** [comment]
To close this subject:
**mixer.clipAction( mesh.geometry.animations[ 1 ] ).play** is NOT a function.
So you cannot use **playMe1()**

**Comment by knod** [comment]
Then why does this work? `this.playMe2 = function() { mixer.clipAction( mesh.geometry.animations[ 1 ] ).play() };`

`.play()` is being used as a function.

**Comment by knod** [comment]
No snark intended, just trying to understand the difference.

**Comment by mrdumb** [comment]
It works because
**mixer.clipAction( mesh.geometry.animations[ 1 ] ).play()**
and
**this.playMe2 = function() { ... }**
are both functions.

**Comment by knod** [comment]
I don't understand your explanation. Maybe you can make a fiddle that replicates the error in the way you describe, show how that behavior makes sense. If we want to talk about it more, though, we should take this into an irc channel or something because this is looking more like a conversation.

**Comment by mrdumb** [comment]
**play() = function()**
**play = variable** (attribute, property, whatever)

**Comment by knod** [comment]
> mixer.clipAction( mesh.geometry.animations[ 1 ] ).play is NOT a function.
> So you cannot use playMe1()

https://jsfiddle.net/4dag9ny6/ `myObject.playMe1();` works just fine.

**Comment by mrdumb** [comment]
> https://jsfiddle.net/4dag9ny6/ myObject.playMe1(); works just fine. 

Your example and three.js are different things.
I have to go now.

**Comment by weiserhei** [comment]
I have fixed the jsfiddle here, use Left or Right mouse button to play the model animation. (Right Mouse throws for me, thats the issue of this thread)

http://jsfiddle.net/f17Lz5ux/5364/

**Comment by mrdoob** [comment]
It boils down to this:

```js
mesh.userData.playMe1 = run.play; 
mesh.userData.playMe2 = function() { run.play() };
```
The reason `playMe2` doesn't work is because `run.play()` is being executed in a different context. `.play()` probably expects to be executed in a specific context and is looking for `this._mixer`. The current implementation is delicate like that.

I hope to have a look at all that code eventually and make it more robust.

**Comment by mrdoob** [comment]
It boils down to this:

```js
mesh.userData.playMe1 = run.play; 
mesh.userData.playMe2 = function() { run.play() };
```

It's all about the context `play()` is being executed. 

The `playMe2` approach results into `run.play()`. `play()` is being executed in the `run` context. `run` has a `_mixer` property.

The `playMe1` approach results into `userData.play()`. `play()` is being executed in the `userData` context. `userData` does not have a `_mixer` property.

As I mentioned over IRC, this is a Javascript idiosyncrasy. We all bumped into this eventually.

**Comment by mrdumb** [comment]
> The reason playMe1 doesn't work is because run.play() is being executed in a different context. 

Not only that.
Logically speaking, the "playMe1" approach is not correct.
